### PR TITLE
put buf Pool back

### PR DESCRIPTION
**Pull request by jiuker**
put buf Pool back

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Unit tests added/updated
- [ ] Internal documentation updated
- [ ] Create a documentation update request [here](https://github.com/minio/docs/issues/new?label=doc-change,title=Doc+Updated+Needed+For+PR+github.com%2fminio%2fminio%2fpull%2fNNNNN)

### CODE REVIEW

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

File: internal/s3select/select.go [hunk 1]
```diff
@@ -546,6 +546,7 @@ func (s3Select *S3Select) Evaluate(w http.ResponseWriter) {
 			return false
 		}
 		outputQueue = outputQueue[:0]
+		bufPool.Put(buf)
```
Comment (harshavardhana) [inline internal/s3select/select.go]: We need to add `defer bufPool.Put(buf)` in the beginning and remove all other places before each return.

```diff
diff --git a/internal/s3select/select.go b/internal/s3select/select.go
index 900cc72c5..be93a79f3 100644
--- a/internal/s3select/select.go
+++ b/internal/s3select/select.go
@@ -522,6 +522,7 @@ func (s3Select *S3Select) Evaluate(w http.ResponseWriter) {
        sendRecord := func() bool {
                buf := bufPool.Get().(*bytes.Buffer)
                buf.Reset()
+               defer bufPool.Put(buf)
 
                for _, outputRecord := range outputQueue {
                        if outputRecord == nil {
@@ -529,12 +530,10 @@ func (s3Select *S3Select) Evaluate(w http.ResponseWriter) {
                        }
                        before := buf.Len()
                        if err = s3Select.marshal(buf, outputRecord); err != nil {
-                               bufPool.Put(buf)
                                return false
                        }
                        if buf.Len()-before > maxRecordSize {
                                writer.FinishWithError("OverMaxRecordSize", "The length of a record in the input or result is greater than maxCharsPerRecord of 1 MB.")
-                               bufPool.Put(buf)
                                return false
                        }
                }
@@ -542,7 +541,6 @@ func (s3Select *S3Select) Evaluate(w http.ResponseWriter) {
                if err = writer.SendRecord(buf); err != nil {
                        // FIXME: log this error.
                        err = nil
-                       bufPool.Put(buf)
                        return false
                }
                outputQueue = outputQueue[:0]
```

This is the correct fix.

**Review by harshavardhana — CHANGES_REQUESTED** [review]

**Comment by jiuker** [comment]
@harshavardhana review plz

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

File: internal/s3select/select.go [hunk 2]
```diff
@@ -522,27 +522,24 @@ func (s3Select *S3Select) Evaluate(w http.ResponseWriter) {
 	sendRecord := func() bool {
 		buf := bufPool.Get().(*bytes.Buffer)
 		buf.Reset()
-
+		defer bufPool.Put(buf)
```
Comment (klauspost) [inline internal/s3select/select.go:525]: No. The buffer is handed over to the writer and cannot be reused.

The buffer is properly reused here: https://github.com/minio/minio/blob/d4e0f13bb3cde2db6e39c620f8e8f2bc37d99cf0/internal/s3select/message.go#L333

Comment (jiuker) [reply internal/s3select/select.go:525]: O,yeah

**Pull request closed without merging**